python/tests/test_ownership_dataset.py

```diff
@@ -0,0 +1,275 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+# Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+# @author Jan-Lukas Wynen
+from copy import copy, deepcopy
+
+import pytest
+import scipp as sc
+
+
+def make_data_array():
+    v = sc.array(dims=['x'], values=[10, 20], unit='m')
+    c = sc.array(dims=['x'], values=[1, 2], unit='s')
+    a = sc.array(dims=['x'], values=[100, 200])
+    m = sc.array(dims=['x'], values=[True, False])
+    da = sc.DataArray(v, coords={'x': c}, attrs={'a': a}, masks={'m': m})
+    return da, v, c, a, m
+
+
+def test_own_darr_set():
+    # Data and metadata are shared
+    da, v, c, a, m = make_data_array()
+    da['x', 0] = -10
+    da.data['x', 1] = -20
+    da.coords['x']['x', 0] = -1
+    da.attrs['a']['x', 0] = -100
+    da.masks['m']['x', 0] = False
+    c['x', 1] = -2
+    a['x', 1] = -200
+    m['x', 1] = True
+    da.unit = 'kg'
+    da.coords['x'].unit = 'J'
+    assert sc.identical(
+        da,
+        sc.DataArray(
+            sc.array(dims=['x'], values=[-10, -20], unit='kg'),
+            coords={'x': sc.array(dims=['x'], values=[-1, -2], unit='J')},
+            attrs={'a': sc.array(dims=['x'], values=[-100, -200])},
+            masks={'m': sc.array(dims=['x'], values=[False, True])}))
+    assert sc.identical(v, sc.array(dims=['x'], values=[-10, -20], unit='kg'))
+    assert sc.identical(c, sc.array(dims=['x'], values=[-1, -2], unit='J'))
+    assert sc.identical(a, sc.array(dims=['x'], values=[-100, -200]))
+    assert sc.identical(m, sc.array(dims=['x'], values=[False, True]))
+
+    # TODO intentional?
+    # Assignments overwrite data but not metadata.
```

Comment:
Yes, intentional, because it is not an assignment, but a `__setitem__` call on a dict?